gdal/port/cpl_vsil_hdfs.cpp

```diff
@@ -120,17 +122,28 @@ int VSIHdfsHandle::Seek(vsi_l_offset nOffset, int nWhence)
   }
 }
 
-vsi_l_offset VSIHdfsHandle::Tell()
+vsi_l_offset
+VSIHdfsHandle::Tell()
 {
   return hdfsTell(poFilesystem, poFile);
 }
 
-size_t VSIHdfsHandle::Read(void *pBuffer, size_t nSize, size_t nMemb)
+size_t
+VSIHdfsHandle::Read(void *pBuffer, size_t nSize, size_t nMemb)
 {
-  return hdfsRead(poFilesystem, poFile, pBuffer, nSize * nMemb);
+  int bytes = hdfsRead(poFilesystem, poFile, pBuffer, nSize * nMemb);
+  if (bytes == 0) {
+    bEOF = true;
```

Comment:
I changed `Read` to have the following behavior:

> The `Read` function returns a non-zero value if and only if the read
> from HDFS succeeded.  Given a read request of more than zero bytes,
> `hdfsRead` returns zero if and only if EOF has been reached, so the EOF
> state is set to be true if that is the case.  `hdfsRead` returns -1 if
> and only if the read failed, in that case `Read` returns zero with the
> EOF state cleared.

The `hdfsRead` function returns zero (for a non-zero requested range) if and only if EOF has been reached:
```
@return      On success, a positive number indicating how many bytes
             were read.
             On end-of-file, 0.
             On error, -1.  Errno will be set to the error code.
             Just like the POSIX read function, hdfsRead will return -1
             and set errno to EINTR if data is temporarily unavailable,
             but we are not yet at the end of the file.
```
so I think that the behavior should now be correct. `errno` is also properly set after this code has run in case that is used anywhere.